src/chocl8.zig

```diff
@@ -0,0 +1,303 @@
+const std = @import("std");
+
+var prng = std.rand.DefaultPrng.init(0);
+
+const font = [_]u8{
+    0xF0, 0x90, 0x90, 0x90, 0xF0, // 0
+    0x20, 0x60, 0x20, 0x20, 0x70, // 1
+    0xF0, 0x10, 0xF0, 0x80, 0xF0, // 2
+    0xF0, 0x10, 0xF0, 0x10, 0xF0, // 3
+    0x90, 0x90, 0xF0, 0x10, 0x10, // 4
+    0xF0, 0x80, 0xF0, 0x10, 0xF0, // 5
+    0xF0, 0x80, 0xF0, 0x90, 0xF0, // 6
+    0xF0, 0x10, 0x20, 0x40, 0x40, // 7
+    0xF0, 0x90, 0xF0, 0x90, 0xF0, // 8
+    0xF0, 0x90, 0xF0, 0x10, 0xF0, // 9
+    0xF0, 0x90, 0xF0, 0x90, 0x90, // A
+    0xE0, 0x90, 0xE0, 0x90, 0xE0, // B
+    0xF0, 0x80, 0x80, 0x80, 0xF0, // C
+    0xE0, 0x90, 0x90, 0x90, 0xE0, // D
+    0xF0, 0x80, 0xF0, 0x80, 0xF0, // E
+    0xF0, 0x80, 0xF0, 0x80, 0x80, // F
+};
+
+/// instructions_per_cycle is the number of instructions per a cycle at 60Hz
+pub const instructions_per_cycle = 12;
+
+const Instruction = struct {
+    short: u16,
+    w: u4,
+    x: u4,
+    y: u4,
+    n: u4,
+    nn: u8,
+    nnn: u12,
+
+    fn decode(lo: u8, hi: u8) Instruction {
+        return Instruction{
+            .short = @as(u16, hi) << 8 | @as(u16, lo),
+            .w = @truncate(u4, hi >> 4),
+            .x = @truncate(u4, hi),
+            .y = @truncate(u4, lo >> 4),
+            .n = @truncate(u4, lo),
+            .nn = lo,
+            .nnn = @as(u12, hi & 0xF) << 8 | @as(u12, lo),
+        };
+    }
+};
+
+const Chip8 = struct {
+    /// mem is 4096 bytes of memory
+    /// 0x0 to 0x1FF is reserved for internal use
+    /// 0x200 is the entry point at which a ROM should be loaded
+    mem: [4096]u8 = [_]u8{0} ** 4096,
+    /// stack is a stack of size 16 for tracking mem locations
+    stack: [16]u16 = [_]u16{0} ** 16,
+    /// sp is the stack pointer
+    sp: u8 = 0,
+    // pc is the program counter
+    pc: u16 = 0,
+    /// index is a register for pointing at mem locations
+    index: u16 = 0,
+    /// delay is a counter decremented at 60hz
+    delay: u8 = 0,
+    /// sound is a counter decremented at 60hz and produces a tone when sound > 0
+    sound: u8 = 0,
+    /// reg is a 16 wide bank of 8-bit registers
+    reg: [16]u8 = [_]u8{0} ** 16,
+    // fbuf is the framebuffer of the screen
+    fbuf: [32]u64 = [_]u64{0} ** 32,
+    // counter tracks the number of instructions executed
+    counter: u64 = 0,
+
+    fn init() Chip8 {
+        var c8 = Chip8{};
+        //@memset(@ptrCast([*]u8, &c8), 0, @sizeOf(Chip8));
+        @memcpy(c8.mem[0x50..], font[0..], font.len);
+        return c8;
+    }
+
+    pub fn loadRom(self: *Chip8, data: []const u8) void {
+        // validate size will fit in memory
+        if (self.mem.len - 0x200 < data.len) {
+            // return error
+        }
+        @memcpy(self.mem[0x200..], @ptrCast([*]const u8, data[0..]), data.len);
```

Comment:
The `@ptrCast` and `[0..]` here is superfluous: all `[n0..n1?]` does is get a subslice, and simply taking the `.ptr` field of the slice is sufficient. In fact, although idk iirc, but I believe `[]const T` coerces to `[*]const T`, meaning even that might be superfluous. Give the latter a try, do the former if it doesn't.

Reply:
This also goes for every other apparently superfluous instance of slicing and `@ptrCast` of a slice.